dataset.py

```
@@ -27,23 +27,205 @@ def __reduce_ex__(self, proto):
-        list.append(self, self.materialize(x))
+class DataIterator():
+    """
+    Build an iterator for Data used.
+    """
+    def __init__(self, data_controller):
+        self.__data_controller = data_controller
+        self.__index = -1
 
-    def extend(self, iterable):
-        list.extend(self, list(map(self.materialize, iterable)))
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.__index + 1 < len(self.__data_controller):
+            self.__index += 1
+            return self.__data_controller[self.__index - len(self.__data_controller)]
+        self.__index = -1
+        raise StopIteration
+
+class DataController():
+    """
+    Control the data privacy.
+    """
+    def __init__(self, data, key=None):
+        self.__data = data
+        self.__key = key
+        self.__end = len(self.__data)
+        self.today = data[0].date if len(self.__data) != 0 else None
+        self.workdays = None
+
+    def __getitem__(self, index):
+        """
+        Use self.__index to contol the data privacy.
+        """
+        if isinstance(index, slice):
+            if index.step == 0:
+                raise IndexError("slice step can't be zero")
+            start = index.start
+            stop = index.stop
+            step = 1 if index.step is None else index.step
+            if start is None:
+                start = -self.__end if step > 0 else -1
+            if stop is None:
+                stop = 0 if step > 0 else -self.__end - 1
+            if start >= 0 or stop > 0:
+                raise IndexError("list index out of range")
+            return DataController(
+                [self[x] for x in range(start, stop, step) if self[x] is not None],
+                self.__key
+            )
+        if index >= 0:
+            raise IndexError("list index out of range")
+        if self.__end + index < 0:
+            if self.workdays is None or len(self.workdays) + index < 0:
+                return None
```

Can we raise an exception here? This is very fatal because we won't be able to simulate alphas in such cases.

---

That means if the date which writer required was older than our workdays, we should decline the require and maybe raise some error ?